keps/sig-node/2727-grpc-probe/README.md

```diff
@@ -0,0 +1,168 @@
+# KEP-2727: Add GRPC Probe
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Goals](#goals)
+- [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Alpha](#alpha)
+    - [Beta](#beta)
+    - [GA](#ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+- [References](#references)
+<!-- /toc -->
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement issue in release milestone, which links to KEP dir in
+  [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] KEP approvers have approved the KEP status as `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture
+      and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in
+  [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents,
+  links to mailing list discussions/SIG meetings, relevant PRs/issues,
+  release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Goals
+
+Enable grpc probe natively from Kubelet without requiring users to package a
+gRPC healthcheck binary with their container.
+
+https://github.com/grpc-ecosystem/grpc-health-probe
+https://github.com/grpc/grpc/blob/master/doc/health-checking.md
+
+## Non-Goals
+
+Add gRPC support in other areas of K8s (e.g. Services).
+
+## Proposal
+
+Host: pod ip
+
+```yaml
+    readinessProbe:
+      grpc:                     #+
+        port: 9090              #+
+        service: my-service     #+
+      initialDelaySeconds: 5
+      periodSeconds: 10
+```
+
+This healthcheck request will be identified with the following gRPC
+`User-Agent` metadata:
+
+```
+User-Agent: kube-probe/K8S_MAJOR_VER.K8S_MINOR_VER
+```
+
+Example:
+
+```
+User-Agent: kube-probe/1.22
+```
+
+### Risks and Mitigations
+
+1. Adds more code to Kubelet and surface area to Pod.Spec. *Response*: we
+   expect that this will be generally useful given broad gRPC adoption in the
+   industry.
+
+## Design Details
+
+```go
+// core/v1/types.go
+
+type Handler struct {
```

Comment:
this structure is shared with prestop and prestart handlers. We already have mismatch of what is supported for probes vs. hooks - sockets make no sense for hooks.

But it would be nice to support grpc hooks, even if the protocol will be the same as the health check

Reply:
What method should the hooks call? And with what arguments? I'm assuming you mean to say that we don't just ping the healthcheck service.method.

Reply:
that what I meant, simplest approach will be to say that hooks will implement the same protocol. My biggest worry is a confusion when authoring hooks - gRPC is defined, but not supported. Secondly, it will be just a handy think for grpc-only services

Reply:
discussed in a chat: will leave this out of scope for now, as we need to understand:

* what service.method makes sense for the hooks. healthcheck makes sense for pre-start, but it is kind of a mismatch for stop hook